[BugFix] Fix NPE in Iceberg metadata table query due to missing Configuration propagation

j1wonpark opened the pull request:
## Why I'm doing:

When querying Iceberg metadata tables (history, snapshots, refs, etc.),
NPE occurs with message: "Cannot invoke Configuration.size() because conf is null".

This happens because:
1. `IcebergCachingFileIO.getWrappedIO()` returns `ResolvingFileIO` without Configuration set
2. When `SerializableTable` serializes the FileIO, it calls `serializeConfWith()`
3. `SerializableConfSupplier` constructor calls `conf.size()` on null Configuration
  
## What I'm doing:

Fix `getWrappedIO()` to propagate Configuration to wrappedIO before returning.
- If `conf` is set via `setConf()`, use it
- If `conf` is null (e.g., REST catalog), use default `new Configuration()`


## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.0
  - [x] 3.5
  - [ ] 3.4
  - [ ] 3.3


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Fixes missing Hadoop `Configuration` propagation to the underlying `ResolvingFileIO`, preventing NPE during metadata table operations.
> 
> - In `IcebergCachingFileIO.initialize`, if `conf` is already set, forward it to `wrappedIO`
> - In `setConf`, also forward the config to `wrappedIO` when it exists
> - Adds `testWrappedIOConfigurationPropagation` to verify `HadoopConfigurable.serializeConfWith` works and no NPE occurs
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bb2649854f40a618a9c616ba02e4af3956db90f9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

j1wonpark committed:
[Test] Add test cases for IcebergCachingFileIO configuration propagation

Signed-off-by: jiwon <jpark92@outlook.kr>

j1wonpark committed:
[BugFix] Fix NPE in Iceberg metadata table query due to missing Configuration propagation

Signed-off-by: jiwon <jpark92@outlook.kr>

j1wonpark committed:
[Refactor] Address SonarCloud code smells

Signed-off-by: jiwon <jpark92@outlook.kr>

alvin-celerdata commented:
@cursor review

Youngwb commented on fe/fe-core/src/main/java/com/starrocks/connector/iceberg/io/IcebergCachingFileIO.java:
```diff
@@ -196,6 +196,8 @@ public void deleteFile(String path) {
     }
 
     public FileIO getWrappedIO() {
+        Configuration configuration = conf != null ? conf.get() : new Configuration();
```
you can use `wrappedIO.setConf(buildConfFromProperties(properties, path));`  I'm curious about why newInputFile wasn't used before getConf. Here, the conf is set, and before reading the table metadata, newInputFile should be used.

Youngwb replied:
You can debug the process of querying Iceberg. When parsing an Iceberg table, it uses newInputFileto read remote metadata files. Querying Iceberg metadata tables also follows this process, so newInputFile will indeed be invoked; otherwise, Iceberg metadata tables using vending credentials would also encounter issues.

j1wonpark replied:
Thanks for the insightful feedback! After digging deeper, I found the real issue.

You're right that `newInputFile()` sets the conf to `wrappedIO`. But here's the thing - in the metadata table query path, `newInputFile()` is never called.

Looking at the stack trace:

```
NPE: Cannot invoke "Configuration.size()" because "conf" is null
    at SerializableTable$SerializableConfSupplier.<init>
    at ResolvingFileIO.serializeConfWith
    at SerializableTable.fileIO
    at SerializableTable.<init>
    at IcebergMetadata.getSerializedMetaSpec  ← starts here
```

This path is for queries like `SELECT * FROM iceberg_table$files`. Here we're just serializing the table to send to BE - no actual file reading happens at this point. So `newInputFile()` never gets a chance to set the conf.

The flow is:
```
getSerializedMetaSpec()
  → getWrappedIO()        // get ResolvingFileIO
  → new SerializableTable // serialize for BE
    → serializeConfWith() // conf is null!
```

Also, I found that Iceberg's `CatalogUtil.loadFileIO()` calls `setConf()` **before** `initialize()`. So when `setConf()` is called, `wrappedIO` doesn't exist yet.

The fix handles both cases:
- `setConf()`: store conf, propagate to wrappedIO if it exists
- `initialize()`: after creating wrappedIO, apply the stored conf

This way, `wrappedIO.conf` is always properly set regardless of call order or code path.

j1wonpark committed:
[BugFix] Fix NPE in IcebergCachingFileIO when setConf called before initialize

Iceberg's CatalogUtil.loadFileIO() calls setConf() before initialize(),
causing NPE when wrappedIO.setConf() is called on null wrappedIO.
- Add null check in setConf() for wrappedIO
- Apply stored conf to wrappedIO in initialize()
- Remove lazy conf setting from getWrappedIO()
- Remove invalid test case for setConf-less scenario

Signed-off-by: j1wonpark <jpark92@outlook.kr>
